### PR TITLE
Update to Scala Dom Types v17.0.0

### PR DESCRIPTION
**Pull request by 2chilled**
Currently there is almost no support for [HTML Global Attributes](https://www.w3schools.com/tags/ref_standardattributes.asp) in html dsl. I quickly hacked together support for data-*, but I think it makes a lot of sense to support all of them properly. WDYT?

Also, what's the plan for fs2.dom types and their methods? Typecasting to underlying scalajs-dom types?

### CODE REVIEW

**Comment by armanbilge** [comment]
> Currently there is almost no support for [HTML Global Attributes](https://www.w3schools.com/tags/ref_standardattributes.asp)

Hmm, I see. Are they missing from scala-dom-types? Would it make sense to add them there and implement them here?

> Also, what's the plan for fs2.dom types and their methods? Typecasting to underlying scalajs-dom types?

Sorry, what's the question? This branch is already using fs2.dom types.

File: calico/src/main/scala/calico/html.scala
Comment (armanbilge) [inline calico/src/main/scala/calico/html.scala:507]: I'm not sure if this is defined correctly. Compare with `ClassProp`.
https://github.com/armanbilge/calico/blob/c96718ba1e5f63c54837aa9a78772059040aa71a/calico/src/main/scala/calico/html.scala#L474-L475

**Comment by armanbilge** [comment]
> One could get rid of casting by doing the wrapping within fs2-dom.

Hmm, I'm not sure what you mean.

The reason I use casting there (and many other places) is for performance. The problem is that a `given` with type parameters compiles to a `def`. So each time you use it, you would be allocating a new `Modifier`.

Instead, I allocate the modifier exactly once as a `val _forStringSignal`. Then, the `inline given` can always return the same instance instead of allocating a new one, but it has to use a cast. This works due to type erasure.

File: calico/src/main/scala/calico/html.scala
Comment (2chilled) [reply calico/src/main/scala/calico/html.scala:507]: > I'm not sure if this is defined correctly. Compare with `ClassProp`.

Hm, indeed I guess there's the property name needed instead of attribute name. That'd be `dataset`. Interestingly it works nonetheless in my simple use case.

**Comment by armanbilge** [comment]
> The implementation is in dom, not in fs2.dom.

Ah I see. In this case, that is also a performance optimization: `e.appendChild(dom.document.createTextNode(head))` can be done in a single `F.delay(...)` instead of several chained effects if we used fs2-dom wrapper APIs.

> I'm wondering what the strategy for fs2.dom is

Fair question. I would say that everything qualifies. Ideally, fs2-dom would wrap the entire DOM API safely. The challenges with this are:
1. the entire DOM API is huge
2. it's not always obvious the best API design to use
3. there is a performance cost for wrappers

That's why I'm currently following a "lazy" approach, where I add new wrapper APIs based on practical needs and use. I think `appendChild`, `createTextNode`, and `textContent` would all be good additions.

File: calico/src/main/scala/calico/html.scala
Comment (armanbilge) [reply calico/src/main/scala/calico/html.scala]: Actually, I was confused about the type parameter: `DataProp.SingleConstantModifier`.

Comment (armanbilge) [reply calico/src/main/scala/calico/html.scala]: Ahhh, interesting! Do you have an example how it's used?

Comment (2chilled) [reply calico/src/main/scala/calico/html.scala]: Sure: ```
          div(
            data("bsToggle") := "collapse",
            cls := "todoapp",
```
Attribute view: `<div class="todoapp" data-bs-toggle="collapse"`. Example is nonsensical but it should show the point.

Comment (armanbilge) [reply calico/src/main/scala/calico/html.scala]: Thanks, that's helpful! So then I think I'm confused why `DataProp` extends `Prop`, since it defines its own `:=` ?

Comment (2chilled) [reply calico/src/main/scala/calico/html.scala]: > Thanks, that's helpful! So then I think I'm confused why `DataProp` extends `Prop`, since it defines its own `:=` ?

That's what I meant by "quickly hacked together" ;) I think we can reuse most functionality for `DataProp` from `Prop` by just abstracting a bit over `setProp`. Will PR this tomorrow.

Comment (mahe-ymc) [reply calico/src/main/scala/calico/html.scala]: Should look a bit better now

Comment (2chilled) [reply calico/src/main/scala/calico/html.scala]: Should look better now

**Comment by 2chilled** [comment]
@armanbilge I started implementing an App based on this PR. Things are going quite well, but I noticed one strange behavior which I want to share with you, hoping you can spot something obvious. The minimized example is this:

```
import calico.html.Html
import calico.router.Router
import calico.router.Routes
import cats.effect.Concurrent
import cats.effect.kernel.Ref
import cats.effect.syntax.*
import cats.effect.syntax.all.*
import cats.Monad
import cats.syntax.all.*
import fs2.concurrent.SignallingRef
import org.http4s.*
import org.http4s.syntax.all.*

object ReloadBug:
  def apply[F[_]: Concurrent: Html: Monad: Router] =
    val html = summon[Html[F]]
    import html.{*, given}

    for {
      sigRef <- SignallingRef[F]
        .of[String]("init")
        .toResource
      routes <- routes(sigRef).toResource
      d <- div(
        cls <-- sigRef.map(_.pure[List]),
        summon[Router[F]].dispatch(routes),
        a(
          "Go to reloadbug2",
          onClick --> (_.foreach(_ => summon[Router[F]].navigate(uri"/reloadbug2")))
        ),
        a(
          "Go to reloadbug",
          onClick --> (_.foreach(_ => summon[Router[F]].navigate(uri"/reloadbug")))
        )
      )
    } yield d

  private def routes[F[_]: Concurrent: Html](ref: Ref[F, String]) =
    val html = summon[Html[F]]
    import html.{*, given}

    val r1 = Routes.one[F] {
      case u if u.path === uri"/reloadbug".path =>
        ()
    }(_ => ref.set("Set by reloadbug route").toResource >> h1("ReloadBug"))

    val r2 = Routes.one[F] {
      case u if u.path === uri"/reloadbug2".path =>
        ()
    }(_ => ref.set("Set by reloadbug2 route").toResource >> h1("ReloadBug"))

    (r1, r2).mapN(_ |+| _)
```

Once /reloadbug is initially loaded, div class is "init". I expected it to be "Set by reloadbug route". When clicking on the Links, div class is set by the corresponding routes as expected. What I know is, the signal change on initial load can be seen by `<--` call, but somehow it doesn't get propagated to the actual dom element. Do you have an idea why this is? If not, I'll look deeper tomorrow.

**Comment by armanbilge** [comment]
> I started implementing an App based on this PR. Things are going quite well

Awesome, great to hear!

> but I noticed one strange behavior which I want to share with you, hoping you can spot something obvious.

Hmm, that does seem strange. Unfortunately I do not see anything obvious either, sorry 🤔 

> What I know is, the signal change on initial load can be seen by `<--` call

How do you know this, did you add some sort of debug `println` ?

---

Btw, sorry things have been slow here. scala-dom-types is undergoing a major redesign in https://github.com/raquo/scala-dom-types/pull/89 so we have a big upgrade coming up 😅

**Comment by 2chilled** [comment]
@armanbilge Even more strange, https://github.com/armanbilge/calico/pull/117/commits/ad758c247e438fb38f529f2c8a0a521246f7aa66 is a workaround..

File: calico/src/main/scala/calico/syntax.scala
Comment (armanbilge) [inline calico/src/main/scala/calico/syntax.scala:71]: Ah I see! Probably this hack is broken somehow 🤔

**Comment by 2chilled** [comment]
@armanbilge First of all, happy new year :) 

Next missing piece for me are SVG tags and attrs. Should I add them in the same way as I did for aria? This time however I'll also need to add some new Wrapper types to fs2-dom.

**Comment by armanbilge** [comment]
@2chilled happy new year! :)

Yes, you can start with fs2-dom. `SvgElement` will also have to be an opaque type since is in the same inheritance with `HtmlElement`.

For changes to Calico I feel a bit anxious until we can update to the new scala-dom-types. Have you had a chance to look at this PR?
- https://github.com/raquo/scala-dom-types/pull/89

If you have time, it would be good to experiment with how we will use that in Calico.

**Comment by 2chilled** [comment]
> SvgElement will also have to be an opaque type since is in the same inheritance with HtmlElement.

:+1: 

> For changes to Calico I feel a bit anxious until we can update to the new scala-dom-types. Have you had a chance to look at this PR?

I didn't look into details yet, but my hope is that we almost get a source compatible API to the current one. Difference is just that API is generated instead of hand-written, with the resulting "benefit" that there are less type params. But I can spend 1 or 2 hours tomorrow trying to lift 0.2-redesign branch into current SDT codegen SNAPSHOT, then we'll hopefully have more facts at hand.

**Comment by armanbilge** [comment]
> but my hope is that we almost get a source compatible API to the current one. Difference is just that API is generated instead of hand-written, with the resulting "benefit" that there are less type params.

Yes, exactly, it should be entirely/nearly source-compatible :)

Since we have control over source-generation now, we may be able to do some additional improvements e.g. use Scala 3 `inline` modifiers. But we can explore that in a later step.

**Comment by 2chilled** [comment]
@armanbilge I gave codegen a shot, see https://github.com/2chilled/calico/tree/pr/0.2-redesign-codegeneration

Unfortunately codegen cannot be used as is because the generators are not expressive enough in regards to kinds of generated type constructors. For example, to generate the constructor of `HtmlTag` the following generator is used:

```
s"def ${keyImplName}[$scalaJsElementTypeParam <: $baseScalaJsHtmlElementType](key: String, void: Boolean = false): ${keyKind}[$scalaJsElementTypeParam] = ${keyKindConstructor(keyKind)}(key, void)"
```
Obviously the kind of `keyKind = HtmlTag` is fixed to `* -> *`, which does not match our current calico.html.HtmlTag kind, which is `(* -> *) -> * -> *`, `(* -> *)` being the generic "effect type" `F[_]`. 

Therefore we need to either (1) make the SDT generators more expressive, or (2) try to push the effect types away from data definitions but into behavior (function) definitions. I guess (2) isn't really an option because we'd need to replace fs2-dom opaque types, since for these the effect types are necessary at data definition level. And without opaque types there is again allocation cost to pay, which we surely want to avoid.

As (1) seems to be the only option, let's look a bit deeper. Instead of generating something like: 

```
trait HtmlTags {
  def htmlTag[Ref <: dom.html.Element](key: String, void: Boolean = false): HtmlTag[Ref] = new HtmlTag(key, void)
```

We'd need to be able to instruct the generator to generate something like this: 

```
trait HtmlTags[F[_]] {
  def htmlTag[Ref <: fs2.dom.Element[F]](key: String, void: Boolean = false)(implicit F: Async[F]): HtmlTag[F, Ref] = new HtmlTag(key, void)
```
Note that I think we cannot use Scala 3 `trait` params, since AFAIK SDT understandably still wants to support Scala 2.

Last but not least this would have to be synced with @raquo, the maintainer of SDT. One could certainly keep the codegen API as it is today, and offer the effectful one separately, as I think the majority of SDT users won't have an interest to support first class effects. But even when @raquo agrees, this will be some work to implement.

**Comment by armanbilge** [comment]
@2chilled thank you very much for investigating that!

I thought we could keep a similar pattern by defining a `protected` method, which the super `trait`s could call without needing to know about `Async`.
https://github.com/armanbilge/calico/blob/671ff27cb87f3ddc73294f00d6b67e591a147d69/calico/src/main/scala/calico/html.scala#L181-L182

--- 

But actually, I think we will need to write our own source-generator in the end anyway. scala-dom-types relies on runtime casting in various places that won't work for our wrapper classes.

**Comment by armanbilge** [comment]
> Do you have an example of such a cast?

Yes, there are two places that the current implementation (Calico 0.1 and old scala-dom-types) rely on casts:

One place is when creating new elements.
https://github.com/armanbilge/calico/blob/1eb3102df2ffac478861dfc5b71685d7fc274676/calico/src/main/scala/calico/dsl.scala#L137

The other place is when listening for events. Actually this cast happens silently in the facade over here.
https://github.com/armanbilge/fs2-dom/blob/28b0cc6c55986cddf2cc9fac4fa1f4bf71bc762e/dom/src/main/scala/fs2/dom/EventTargetHelpers.scala#L32

In both situations, scala-dom-types refines the element/event type to the precise type returned by the underlying JavaScript API. But if we are creating our own wrapper class dependent on the underlying type, then this casting trick will not work.

This won't be a problem for HTML elements so long as we keep using opaque types (since the underlying type is the same). But I think we should use wrapper classes for events, so casting will not work, and we will need to manually override these methods to create the correct wrapper.

**Comment by 2chilled** [comment]
Just rethought that "leave htmlTag abstract" idea. This also won't work, because there's no support for type variables anymore. The whole trick in calico 0.1 to postpone substitution for the concrete `HtmlTag` type relies on this (old) definition: 
```
trait HtmlTagBuilder[T[_ <: DomHtmlElement], -DomHtmlElement] {
  @inline protected def htmlTag[Ref <: DomHtmlElement](tagName: String, void: Boolean): T[Ref]
```
In the codegen Version there is no type var, indeed there isn't even support for introducing one, which can be witnessed here: 

```
trait $traitName $withTraits{$traitThisTypeStr", "}"
```

No type vars means no effect types in data types unfortunately. Without having type var support in SDT generators, I'm afraid we can't use anything from SDT anymore.

**Comment by armanbilge** [comment]
> This also won't work, because there's no support for type variables anymore

@2chilled what about an abstract type for that as well?

```scala
trait Foo {
  protected type Bar
}

class FooImpl extends Foo {
  type Bar = BarImpl
}
```

**Comment by raquo** [comment]
Hey guys, if you want to achieve this with the new SDT:

```
trait HtmlTagBuilder[T[_ <: DomHtmlElement], -DomHtmlElement] {
  @inline protected def htmlTag[Ref <: DomHtmlElement](tagName: String, void: Boolean): T[Ref]
```

Then:

1) Pass `HtmlTagBuilder[T[_ <: DomHtmlElement], -DomHtmlElement]` as `traitName` param to `TagsTraitGenerator`
2) Pass your custom `baseImplDef` to `TagsTraitGenerator`

To clarify, this would mean instantiating `TagsTraitGenerator` yourself instead of using `CanonicalGenerator.generateTagsTrait`. This should take care of generating the trait snippet above, but you might also need other customizations, for which you can subclass `TagsTraitGenerator` (e.g. if you need to override `printDef`). See the suggested strategy for customization in bullet point 4 [here](https://github.com/raquo/scala-dom-types#how-to-use-scala-dom-types-in-your-library).

**Comment by 2chilled** [comment]
@raquo Thanks for joining in! 

> Pass HtmlTagBuilder[T[_ <: DomHtmlElement], -DomHtmlElement] as traitName param
Will try that. Wasn't sure about potential side effects happening when encoding trait name and parameter lists together as trait name, but you surely know your codes :)

**Comment by raquo** [comment]
@2chilled yeah, normally my public APIs are a lot more polished, but I just could not afford to spend the required time to design this code in an elegant or user-friendly way.

I did try to optimize for flexibility though, so despite the API being ugly and at times unintuitive, it's very flexible. Unfortunately you need to look into the code to see what it's doing with all those vars. At least the code is structurally simple, so it's not too hard to figure out what it's doing.

In this case it is indeed fine to put generics in this var, even though the name might imply otherwise.

**Comment by 2chilled** [comment]
@raquo no worries, I think we should all be very thankful for each piece of Open Source Software, no matter the quality. The point is since it's Open Source, everyone has the chance to improve it :)

**Comment by 2chilled** [comment]
https://github.com/2chilled/calico/blob/pr/0.2-redesign-codegeneration/calico/src/main/scala/calico/html/defs/tags/HtmlTagBuilder.scala typechecks

**Comment by 2chilled** [comment]
@armanbilge I have everything except Aria, Svg and Styles in place. Will add Aria tomorrow, then integrate everything into calico.html.Html. When this is done and polished, I will PR against pr/0.2-redesign. Once that is merged I suggest to begin writing smaller Issues and tackle them with smaller PRs. This will enable us to work a bit more focused and overall more organized. Is that ok with you?

**Comment by 2chilled** [comment]
> thanks for all your work on this

Thank you for starting calico. I hope this lib will finally end the way too long history of pain when it comes to GUI programming :laughing: What makes me very optimistic about that is it builds on the very very right fundamentals, which is just awesome. And all of this without the overhead of VDoms. Really appreciate it :+1:

**Comment by armanbilge** [comment]
Woops, merging my other PR closed this one ...

**Comment by 2chilled** [comment]
@armanbilge This is ready for review. I manually tested TODO App and my own App, works fine so far. It's not quite src compatible, some tags have been renamed. All in all very low migration cost for calico users. 

There are some low hanging fruits left, but these require fs2-dom additions. Will PR these separately.

**Comment by armanbilge** [comment]
@2chilled awesome!! Will take a closer look soon, probably not until this weekend.

Question: instead of checking in the generated sources, can we generate them on compile? I think I'd prefer not to check them into source control. I can help set this up as well.

File: .gitignore
Comment (armanbilge) [inline .gitignore:22]: Generated code should go in `sourceManaged`.
https://www.scala-sbt.org/1.x/docs/Howto-Generating-Files.html

File: build.sbt
Comment (armanbilge) [inline build.sbt:28]: sbt has a `sourceGenerators` capability for this.

```scala
Compile / sourceGenerators += <task of type Seq[File]>.taskValue
```

Comment (2chilled) [reply build.sbt:28]: Laminar also has the generated files checked into source control, but I have no idea why.

File: .gitignore
Comment (2chilled) [reply .gitignore:22]: Hmm:
```

Comment (2chilled) [reply .gitignore:22]: Hm, classpath issue is left, scalac doesn't see the classes from sources:
```
[error] 19 |import calico.html.keys.AriaAttr
[error]    |       ^^^^^^^^^^^^^^^^
[error]    |     value keys is not a member of calico.html - did you mean html.defs
```

Comment (2chilled) [reply .gitignore:22]: seems to be a cyclic error, looks like sbt wants to compile sourceManaged in an isolated run. Maybe this can be solved by introducing another sbt subproject. Will try tomorrow.

Comment (armanbilge) [reply .gitignore:22]: Huh, that seems strange. Pretty sure I've used `sourceManaged` with cycles before.

Did you push these changes? I can take a quick look.

Comment (2chilled) [reply .gitignore]: Pushed now

File: build.sbt
Comment (armanbilge) [inline build.sbt:35]: This should be defined in the `.settings()` for the calico subproject, right?

Comment (2chilled) [reply build.sbt:35]: Ahhh, yes, you are perfectly right! Seems to work now. I need some sleep, see you tomorrow :)

File: build.sbt
Comment (raquo) [reply build.sbt]: Laminar has generated sources checked into source control because the fact that they're generated is an implementation detail that the end users don't care about. As a user, I want to see all the code of the library in one place, e.g. I should be able to search the project's github repo to find `idAttr`, `footerTag`, etc. For contributors, the generated sources have a comment on top warning not to edit the files directly, and pointing them to the generator and to SDT. This also essentially makes the generator / SDT an _optional_ dependency, as you can compile Laminar with the generator disabled, if there is ever a need. But mostly I care about UX.

Comment (armanbilge) [reply build.sbt]: > the fact that they're generated is an implementation detail that the end users don't care about

Yeah this is a great perspective and overall I agree. Unfortunately for my users the prospect of checking in generated sources makes me feel queasy 😝

Also I'd argue it's kind of a tooling problem. We upload our source artifacts to maven central, including generated sources. If we can have javadoc.io to browse scaladoc artifacts, then surely we could have a similar service for browsing sources. Also anyone in an IDE would be able to browse the sources too.

Comment (raquo) [reply build.sbt]: It helps if you think of the generator as an ~~unpaid intern~~ coding assistant who pumps out a lot of boring repetitive code, as opposed to a macro that is supposed to create opaque code that nobody ever wants to look into :)

File: calico/src/main/scala/calico/html/Modifier.scala
Comment (armanbilge) [inline calico/src/main/scala/calico/html/Modifier.scala:127]: Did you add these? Calico 0.2 will no longer allow `Stream`-based modifiers, only `Signal`.

Comment (2chilled) [reply calico/src/main/scala/calico/html/Modifier.scala:127]: @armanbilge I guess I added these to make Example.scala compile. Now I've seen that you basically have removed Example.scala. I will revert this but I'm curious why you don't want to support `Stream` modifiers anymore? Obviously `Signal` makes more sense since it asserts that there is a value available from the start. But one potential shortcoming I can see is that there is no `Monad` instance for `Signal`, which makes it not possible to modify the value before the actual DOM elem is modified using another effect like a HTTP request for example. For this use case one has to introduce another `Signal` which gets modified in the background by getting a `Stream` from the initial `Signal`, which seems a bit involved.

Comment (armanbilge) [reply calico/src/main/scala/calico/html/Modifier.scala:127]: > potential shortcoming I can see is that there is no Monad instance for Signal

https://github.com/armanbilge/calico/blob/d0812e8009ea7b390292a0aee0279e23c9d9c302/frp/src/main/scala/calico/frp/frp.scala#L50

Comment (2chilled) [reply calico/src/main/scala/calico/html/Modifier.scala:127]: Shame on me that I didn't know about this. Do you plan to push this upstream?

Comment (2chilled) [reply calico/src/main/scala/calico/html/Modifier.scala]: Removed the Stream instances (again). What about Example.scala? I just modified it to use the `Signal` modifiers instead of `Stream` ones.

Comment (armanbilge) [reply calico/src/main/scala/calico/html/Modifier.scala]: Yes, I've discussed this with Fabio (FS2 maintainer), basically the idea is to "incubate" this instance in Calico, and if it proves to be robust then we can upstream it to FS2.

The implementation and/or test might still be a bit buggy, see:
- https://github.com/armanbilge/calico/issues/101

Comment (2chilled) [reply calico/src/main/scala/calico/html/Modifier.scala]: It just went red after removing the instances

Comment (armanbilge) [reply calico/src/main/scala/calico/html/Modifier.scala]: Yeah, that test is flaking sometimes 😕 so far I have a hard time reproducing it. If you have any insight to that, it would be really great 😁

Comment (armanbilge) [reply calico/src/main/scala/calico/html/Modifier.scala]: Btw a couple more thoughts:

> Now I've seen that you basically have removed Example.scala

Yeah, actually I just use that as a sandbox for testing stuff, probably should rename it. All the examples go in the docs now.

> Obviously `Signal` makes more sense since it asserts that there is a value available from the start.

Yes, I wrote about this in the v0.1.2 release notes. Also it means we can make an optimization: you can retrieve the first value, render the HTML, and then setup the subscriptions in the background. But I still don't have this working perfectly.
https://github.com/armanbilge/calico/releases/tag/v0.1.2

Comment (2chilled) [reply calico/src/main/scala/calico/html/Modifier.scala]: I'll try to reproduce locally

**Comment by 2chilled** [comment]
Facing https://github.com/typelevel/sbt-typelevel/issues/373 here:

```
sbt:root> last calico/Compile/packageSrc
[error] java.util.NoSuchElementException: None.get
[error]         at scala.None$.get(Option.scala:529)
[error]         at scala.None$.get(Option.scala:527)
[error]         at org.typelevel.sbt.TypelevelSettingsPlugin$.$anonfun$perConfigSettings$5(TypelevelSettingsPlugin.scala:284)
[error]         at scala.collection.immutable.List.map(List.scala:293)
[error]         at org.typelevel.sbt.TypelevelSettingsPlugin$.$anonfun$perConfigSettings$4(TypelevelSettingsPlugin.scala:284)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error]         at sbt.Execute.work(Execute.scala:291)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error]         at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
[error]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[error]         at java.base/java.lang.Thread.run(Thread.java:833)
[error] (calico / Compile / packageSrc / mappings) java.util.NoSuchElementException: None.get
```

`show calico/Compile/packageSrc/mappings` also triggers it. 

```
    Compile / packageSrc / mappings := {
      val base = sourceManaged.value
      (Compile / managedSources).value.map(file => file -> file.relativeTo(base).get.getPath)
    },
```
in calico settings is a workaround

**Comment by armanbilge** [comment]
Hmm, last time I saw that issue it was someone else's fault, not sbt-typelevel. Let me take a look.

Btw, I ran out of time yesterday but I'd still like to extract some changes from here as separate PRs. For example, you did some nice re-organizing of `html.scala` into a few files. It's hard to see the diff otherwise.

**Comment by 2chilled** [comment]
> For example, you did some nice re-organizing of html.scala into a few files

Hmm, ok, I can put everything except pure additions back to html.scala

**Comment by armanbilge** [comment]
I'm happy either way, thanks!

File: build.sbt
Comment (armanbilge) [inline build.sbt:65]: Instead of this `PathFinder` trick, can we return the actual list of files generated from the `cachedGenerate` task?

Tangentially, I don't like this caching strategy using `lastScalaDomTypesVersion.txt`. The build already knows the SDT version, and also if we change our code generating sources in `project/` it should be also invalidate the cache. I need to think about a better way to do this.

Comment (mahe-ymc) [reply build.sbt:65]: > Instead of this PathFinder trick, can we return the actual list of files generated from the cachedGenerate task?
Yes we can. The reason I didn't do in the first place is I aimed to stay close to the original SDT API. But yeah, since we needed to overwrite enough generator now, I think this is not a real benefit anymore. On the other side this specific trick is simple enough. But if it bothers you, I can change it.

> I don't like this caching strategy using lastScalaDomTypesVersion.txt
I also don't, for the reasons you mentioned. But at this point where we don't have anything generated checked into source control, it's more or less just a bit inconvenient for calico devs. I think for the full solution we would need to hash sources in project/ together with SDT Version, and save that somewhere. Maybe we can even save it into user tmpdir. Impl would be simple enough.

Comment (armanbilge) [reply build.sbt:65]: > I think for the full solution we would need to hash sources in project/ together with SDT Version, and save that somewhere. Maybe we can even save it into user tmpdir.

The thing is, if the sources in `project/` change, then you have to `reload` the sbt build anyway. So I don't think we need to save this information anywhere at all. Every time you load the build, you have to generate the sources, but that's it, because the only way to change the SDT version or the source generator config is to reload the build.

Comment (2chilled) [reply build.sbt:65]: > Every time you load the build, you have to generate the sources once

... when necessary :) Of course, we can just generate them on load unconditionally. But then, when the rest of the build is modified, in the usual modify/reload loop, sources get needlessly generated again and again. Granted, reloads happen much more seldom than compiles, so this has probably the best cost benefit ratio when it comes to solutions requiring no additional state.

**Comment by 2chilled** [comment]
@armanbilge moved everything back into html.scala to improve the diff.

**Comment by armanbilge** [comment]
Just saw!! Thank you so so much, really appreciate that. Will try and take a proper review pass today.

File: build.sbt
Comment (armanbilge) [reply build.sbt:65]: Yeah, the current scheme is definitely broken. I just did https://github.com/armanbilge/calico/pull/117/commits/1b064b5cc703c266ae60c8ee56b5e9fc32e02de0 and had to manually delete the `lastScalaDomTypesVersion.txt` file which is bad. At the very least it should go into `target/`, so that `clean` makes it go away.

Comment (2chilled) [reply build.sbt:71]: I know, it's bad, totally agree. Added a version without cache but with onload generator. Smallest bottleneck is scalafmt of generated sources, maybe we should just get rid of it to speed reloading up. Although, for me it's fast enough.

Comment (armanbilge) [reply build.sbt:71]: > Smallest bottleneck is scalafmt of generated sources, maybe we should just get rid of it to speed reloading up.

Yes, I am just about to push this ;) completely unnecessary for generated sources. Thanks!

**Comment by armanbilge** [comment]
Waiting for:
- https://github.com/raquo/scala-dom-types/pull/91

**Review by armanbilge — APPROVED** [review]
@2chilled thank you so much for all of your work on this 🙏 really relieved to make this upgrade 😅 

Still there is quite a bit to do, so I created a milestone for v0.2.0. I will open a couple follow-up issues and please make your own!
https://github.com/armanbilge/calico/milestone/2

**Pull request merged**
